### PR TITLE
Reuse failed auth results in AuthN

### PR DESCRIPTION
**Pull request by Kahbazi**

### CODE REVIEW

**Review by davidfowl — APPROVED** [review]
Keep these changes coming 👍🏾

**Review by davidfowl — CHANGES_REQUESTED** [review]
Turns out you can mutate these results 😢

**Comment by Kahbazi** [comment]
> Turns out you can mutate these results 😢

I can't see how! Could you please elaborate? Is it because of the `Exception` object? I would argue that even in the clone method the Exception remains the same instance.

**Review by davidfowl — APPROVED** [review]

File: src/Security/Authentication/Certificate/src/AuthenticateResults.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Authentication.Certificate;
+
+internal static class AuthenticateResults
+{
+    internal static AuthenticateResult NoSelfSigned = AuthenticateResult.Fail("Options do not allow self signed certificates.");
```
Comment (Tratcher) [inline src/Security/Authentication/Certificate/src/AuthenticateResults.cs:8]: Style: Why define these in separate classes if they're only used in one class?

File: src/Security/Authentication/Core/src/HandleRequestResult.cs [hunk 2]
```diff
@@ -8,6 +8,10 @@ namespace Microsoft.AspNetCore.Authentication;
 /// </summary>
 public class HandleRequestResult : AuthenticateResult
 {
+    private static readonly HandleRequestResult _noResult = new() { None = true };
```
Comment (Tratcher) [inline src/Security/Authentication/Core/src/HandleRequestResult.cs:11]: Style: I think we use "NoResult" for class statics, and "_noResult" for class fields.

**Review by Tratcher — COMMENTED** [review]
Can you change AuthenticateResult.NoResult to cache?

There's a theoretical issue with caching AuthenticateResult, it can reference mutable structures like ClaimsPrincipal, AuthenticationProerties, Exception, etc.. NoResult, Skipped, and Handled might be the only cases truly safe to cache because they don't hold any mutable objects. That might be worth overlooking for extremely common scenarios. Are any of these failures especially common?

*edit* Oh, already discussed above 😁.

(Merged since the feedback was trivial and we're at the rc1 deadline)

**Review by Tratcher — APPROVED** [review]

**Comment by Kahbazi** [comment]
> Can you change AuthenticateResult.NoResult to cache?

@Tratcher I've done that in this PR https://github.com/dotnet/aspnetcore/pull/43268

**Pull request merged**